29 - Создание репозитория Справочники

Keithcutt opened the pull request.

Keithcutt committed:
Domain-слой.
Созданы интерактор и репозиторий для работы со справочниками.
Созданы модели.

Keithcutt committed:
Domain-слой.
Созданы интерактор и репозиторий для работы со справочниками.
Созданы модели.

Keithcutt committed:
Data-слой.
Созданы репозиторий и NetworkClient.
Созданы dto-модели, модели запросов и ответов.

Keithcutt committed:
Data-слой.
Созданы репозиторий и NetworkClient.
Созданы dto-модели, модели запросов и ответов.

Keithcutt committed:
Добавлены недостающие имплементации usecase'ов.

Keithcutt committed:
Добавлены недостающие имплементации usecase'ов.

Keithcutt committed:
Добавлен di-модуль

Keithcutt committed:
Добавлен di-модуль

Keithcutt committed:
Обновлены модели в соответствии с обновленным api.
Реализовано кеширование справочников.
Реализовано di.

Keithcutt committed:
Обновлены модели в соответствии с обновленным api.
Реализовано кеширование справочников.
Реализовано di.

Keithcutt committed:
Исправлены мапперы согласно общепринятому код-стайлу.

Keithcutt committed:
Исправлены замечания detekt'a

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/data/dto/CountryDto.kt:
```diff
@@ -0,0 +1,13 @@
+    @SerialName("country_name")
+    val name: String? = null,
+    val cities: List<CityDto>? = null
+)
```
Не думаю что name и cities должны быть nullable, в API такого нет. С бека может не прилететь сразу весь объект, но если прилетит, то все поля будут.

У cities надо тоже прописать `@SerialName("cities")`

Keithcutt replied:
По поводу nulable в api в теле ответа на запрос currencies будет приходить объект типа Country, но с единственным заполненным полем id. Решил не множить модели country чтобы не путаться. 
По поводу аннотации - исправлю.
<img width="336" height="303" alt="image" src="https://github.com/user-attachments/assets/0fc27750-59ce-4134-83b0-99360812b56c" />

nemislimus replied:
Оу...
То есть задублировать пачку Dto это не влом, а одну, значит, модель CountryIdDto(val id: Int) сделать, чтоб потом не мучаться с проверками на null, это как-то не охота... Ну ок :feelsgood:

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/data/localdto/CountryLocalDto.kt:
```diff
@@ -0,0 +1,10 @@
+
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class CountryLocalDto(
+    val id: Int,
+    val name: String? = null,
+    val cities: List<CityLocalDto>? = null
+)
```
А в чём сакральный смысл дублировать Dto модели для локального использования?

Keithcutt replied:
Для того, чтобы при расширении приложения был выбор что мы хотим кешировать, а что не хотим (+ так посоветовал сделать Денис).

nemislimus replied:
Не до конца понял `чтобы при расширении приложения был выбор что мы хотим кешировать, а что не хотим`. 
Имеется в виду - какие то поля возможно нам надо будет убрать из модели и не сохранять или что?
Пока что мы кэшируем модели один в один и я не вижу сценария, где бы мы делали по другому (^_^)

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/domain/model/Currency.kt:
```diff
@@ -0,0 +1,8 @@
+
+data class Currency(
+    val id: Int,
+    val type: String,
```
Для ограниченных типов можно сделать enum. Проще будет потом на UI обрабатывать:
```kotlin
enum class CurrencyType(val currencyValue: String) {
    EUR("EUR"),
    THB("THB"),
    UNKNOWN("UNKNOWN"),
}
```

Keithcutt replied:
В таком случае можно в presentation-слое он будет смотреться уместнее наверное.

Secur1ty-21 replied:
(Не крит) Соглашусь с Виталием, лучше добавить, чтобы видеть в слое бизнес-логики, какими типами ограничены доступные валюты

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/domain/model/Country.kt:
```diff
@@ -0,0 +1,7 @@
+
+data class Country(
+    val id: Int,
+    val name: String? = null,
+    val cities: List<City>? = null
```
Тут тоже не вижу резона использовать nullable...

Keithcutt replied:
Ответил выше

nemislimus replied:
Ну тут то ты можешь в domain модели уже не объект Country внутри Currency использовать, а просто поле с id. При мэпинге из data в domain это надо учесть... 
Просто потом возможно будет не удобно проверять везде на null, много лишнего кода может получиться.

Keithcutt replied:
Окей. Тут тогда уберу nullable. Сделаю поле.

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/data/mapper/ReferenceDataMapper.kt:
```diff
@@ -0,0 +1,138 @@
+    )
+}
+
+@JvmName("mapCountryDtoToDomain")
```
А вот это даже интересно - лайк 👍

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/di/ReferenceDataModule.kt:
```diff
@@ -0,0 +1,57 @@
+import org.koin.dsl.module
+
+val referenceDataModule = module {
+    single { Json { prettyPrint = true } }
```
Json уже в CoreModule был определён.

Keithcutt replied:
Ага, как-то забыл об этом. Уберу.

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/domain/api/ReferenceDataRemoteRepository.kt:
```diff
@@ -0,0 +1,14 @@
+import kotlinx.coroutines.flow.Flow
+
+interface ReferenceDataRemoteRepository {
+    fun getCountries(): Flow<VolleyResult<List<Country>, ErrorType>>
+    fun getCurrencies(): Flow<VolleyResult<List<Currency>, ErrorType>>
+    fun getFaq(): Flow<VolleyResult<Faq, ErrorType>>
```
Тоже думал у себя обернуть в Flow, но в итоге не стал, ибо не очень пока понимаю как мы правильно на UI будем это всё обрабатывать и в state загонять на Compose + посмотрел как у предыдущих было сделано, не стал раздваивать стиль.

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/data/repository/ReferenceDataRemoteRepositoryImpl.kt:
```diff
@@ -0,0 +1,104 @@
+        when (response.isSuccess) {
+            true -> {
+                val result =
+                    (response.body as ReferenceDataResponse.CountriesResponse).countries.mapToDomain()
+                localRepository.saveCountries(result.mapToLocalDto())
+
```
У response body это Any? - через это надо делать безопасное приведение типа и обрабатывать эту ситуацию:
```kotlin
val result =
                    (response.body as? ReferenceDataResponse.CountriesResponse)?.countries?.mapToDomain()
                result?.let {
                    localRepository.saveCountries(result.mapToLocalDto())
                    emit(VolleyResult.Success(result))
                } ?: emit(VolleyResult.Failure(ErrorType.UNKNOWN_ERROR))
```

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/data/repository/ReferenceDataRemoteRepositoryImpl.kt:
```diff
@@ -0,0 +1,104 @@
+            false -> {
+                val countriesFromCache = localRepository.loadCountries()
+
+                if (countriesFromCache != null) {
+                    emit(
+                        VolleyResult.Success(countriesFromCache.mapToDomain())
+                    )
+                } else {
+                    emit(
+                        VolleyResult.Failure(response.resultCode.mapToErrorType())
+                    )
+                }
```
Можно писать проще и чуть красивее:
```kotlin
countriesFromCache?.let {
                    emit(VolleyResult.Success(countriesFromCache.mapToDomain()))
                } ?: emit(VolleyResult.Failure(response.resultCode.mapToErrorType()))
```

Keithcutt replied:
Так и знал, что есть способ получше...

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/data/repository/ReferenceDataRemoteRepositoryImpl.kt:
```diff
@@ -0,0 +1,104 @@
+
+        when (response.isSuccess) {
+            true -> {
+                val result =
+                    (response.body as ReferenceDataResponse.FaqResponse).faqDto.mapToDomain()
```
Та же история как и выше...

nemislimus commented on app/src/main/java/cy/volleybolley/referencedata/data/repository/ReferenceDataRemoteRepositoryImpl.kt:
```diff
@@ -0,0 +1,104 @@
+        when (response.isSuccess) {
+            true -> {
+                val result =
+                    (response.body as ReferenceDataResponse.CurrenciesResponse).currencies.mapToDomain()
+                localRepository.saveCurrencies(result.mapToLocalDto())
```
Тут тоже надо не забыть обработать `as?`

nemislimus reviewed (changes requested):
В целом всё отлично, но как по мне есть моменты которые надо поправить (Аннотации в Dto и обработка результатов в remoteRepo).
Я не очень понял зачем дублированы Dto для localRepo, возможно я чего-то не понимаю))

Keithcutt committed:
Merge branch 'develop' into 29-reference-data-repository-creation

# Conflicts:
#	app/src/main/java/cy/volleybolley/core/di/HttpClientQualifier.kt

nemislimus commented on app/src/main/java/cy/volleybolley/core/DiProvider.kt:
```diff
@@ -3,11 +3,13 @@ package cy.volleybolley.core
+        authViewModelModule,
+        referenceDataModule,
     )
 }
```
накинь плз в список profileModule.

Keithcutt committed:
Исправлены замечания Виталия после первого ревью.
Добавлен profileModule в DiProvider.

Keithcutt committed:
Добавлен модификатор suspend в RemoteReposityory.

Keithcutt committed:
Detekt fix

Keithcutt committed:
Detekt fix 2

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/data/localdto/CityLocalDto.kt:
```diff
@@ -0,0 +1,9 @@
+
+@Serializable
+data class CityLocalDto(
+    val id: Int,
```
(Крит) Тут так же как и в сетевых моделях лучше помечать SerialName, чтобы при изменении наименования переменной не поехала логика сериализации/десериализации

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/data/repository/ReferenceDataLocalRepositoryImpl.kt:
```diff
@@ -0,0 +1,64 @@
+
+    override suspend fun saveCountries(data: List<CountryLocalDto>) {
+        val jsonString = json.encodeToString(data)
+        getFile(COUNTRIES_CACHE).writeText(jsonString)
```
(Крит) При всех операциях взаимодействия с файлом лучше оборачивать их в runCatching, отлавливать ошибку тут и логировать с флагом BuildConfig.DEBUG

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/data/repository/ReferenceDataRemoteRepositoryImpl.kt:
```diff
@@ -0,0 +1,95 @@
+                result?.let {
+                    localRepository.saveCountries(it.mapToLocalDto())
+                    emit(VolleyResult.Success(it))
+                } ?: emit(VolleyResult.Failure(ErrorType.UNKNOWN_ERROR))
```
(Не крит) Для соблюдения ssot после сохранения в файл, надо получить данные оттуда, смаппить в нужную модель и отдать на domain уже то что получили из файла

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/di/ReferenceDataModule.kt:
```diff
@@ -0,0 +1,55 @@
+val referenceDataModule = module {
+    single {
+        val context = get<Context>()
+        context.cacheDir
```
(Не крит) В koin вроде сразу можно получить context
```kotlin
androidContext()
```

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/data/repository/ReferenceDataRemoteRepositoryImpl.kt:
```diff
@@ -0,0 +1,95 @@
+                    )
+                } ?: emit(VolleyResult.Failure(response.resultCode.mapToErrorType()))
+
+            }
```
(Не крит) Думаю эту логику лучше вынести в useCase, больше похоже именно на бизнес логику

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/domain/model/Faq.kt:
```diff
@@ -0,0 +1,5 @@
+package cy.volleybolley.referencedata.domain.model
+
+data class Faq(
```
(Не крит) Когда фишки data class не используются лучше не прописывать его просто так (увеличивает время компиляции и реверс инженерам легче понять логику приложения, т.к data class описывает сам себя дополнительно в аннотации Metadata

Keithcutt replied:
А ты имел ввиду, что проблема только с этим конкретным классом или это касается и других data-классов в моей таске?

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/domain/impl/GetCountriesUseCaseImpl.kt:
```diff
@@ -0,0 +1,15 @@
+
+class GetCountriesUseCaseImpl(private val remoteRepository: ReferenceDataRemoteRepository) :
+    GetCountriesUseCase {
+    override fun execute(): Flow<VolleyResult<List<Country>, ErrorType>> {
```
(Не крит) Думаю  flow в справочниках не очень подходит, т.к. тут мы возвращаем единичный результат, а не наблюдаем за состоянием. Вот тут ответ с галочкой вроде неплохо передает мою мысль -> https://stackoverflow.com/questions/75537679/when-i-should-use-flow-and-when-should-i-use-corountines

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/data/cache/ReferenceDataLocalRepository.kt:
```diff
@@ -0,0 +1,14 @@
+import cy.volleybolley.referencedata.data.localdto.FaqLocalDto
+
+interface ReferenceDataLocalRepository {
+    suspend fun saveCountries(data: List<CountryLocalDto>)
```
(Не крит) Было бы хорошей практикой возвращать результат сохранения (удалось/не удалось)

Secur1ty-21 reviewed (changes requested):
В целом отличная работа, чисто и красиво. Пометил как «крит» то, что может вызвать вылеты или неожиданное поведение в будущем и нужно поправить. То, что «не крит», можно не исправлять и оставить как есть

Keithcutt committed:
Правки после второго ревью.
- Добавлены аннотации к полям LocalDto-классов.
- Data-модификаторы оставлены только у соответствующих классов в domain-слое.
- Все действия c File обернуты в runCatching. Неудачные ошибки будут отображаться в логах (только в debug-версии).
- Методы для кеширования данных справочников теперь возврящают Boolean (успех/неудача).
- Типы валют ограничены одноименным enum-классом.
- Usecase'ы справочников возвращают просто VolleyResult (без flow).
- Изменена логика работы с кешированными справочниках: соблюдается ssot (данные в domain передаются только через localRepository), при неудачном ответе на запрос проверка кеша осуществляется в usecase'е.

Keithcutt committed:
Правки после проверки detekt'а.

Keithcutt committed:
Правки после проверки detekt'а 2.

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/data/repository/ReferenceDataLocalRepositoryImpl.kt:
```diff
@@ -0,0 +1,112 @@
+        }.onFailure { exception ->
+            if (BuildConfig.DEBUG) {
+                Log.e(ERROR_TAG, "Save $COUNTRIES_CACHE error", exception)
+                return false
```
(Не крит) Тут получается что на релизной версии всегда будет true возвращаться

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/data/repository/ReferenceDataLocalRepositoryImpl.kt:
```diff
@@ -0,0 +1,112 @@
+        }.getOrNull()
+    }
+
+    companion object {
```
(Не крит) Можно вот так, тогда не надо для каждого поля писать
```kotlin
private companion object {
        const val COUNTRIES_CACHE = "countries_cache"
        const val CURRENCIES_CACHE = "currencies_cache"
        const val FAQ_CACHE = "faq_cache"
        const val ERROR_TAG = "ReferenceDataCaching"
}
```

Secur1ty-21 commented on app/src/main/java/cy/volleybolley/referencedata/data/repository/ReferenceDataRemoteRepositoryImpl.kt:
```diff
@@ -0,0 +1,86 @@
+                    VolleyResult.Success(
+                        localRepository.loadCountries()?.mapToDomain().orEmpty()
+                    )
+                } ?: VolleyResult.Failure(ErrorType.UNKNOWN_ERROR)
```
(Не крит) Думаю если не получилось сохранить и получить данные , то лучше вернуть Failure с Unknown error, иначе пользователь не сможет выбрать страну и при этом ошибка не покажется

Secur1ty-21 reviewed (approved):
Критических замечаний нет, отличная работа!

Keithcutt committed:
Правки после второго ревью наставника.

Keithcutt committed:
Merge branch 'develop' into 29-reference-data-repository-creation
